### DISCUSSION TITLE
Adding Material Property to Custom Kernel

### DISCUSSION THREAD

**Discussion by alanchavez86** [idaholab/moose#18157] [Q&A General] [answered]
Hi All!

I was wondering if I can get some help with adding a material property to a custom kernel that can compute the magnetic potential in the following equation: 

![image](https://user-images.githubusercontent.com/74686859/123172010-ba5fd880-d439-11eb-99f8-c6c0a2f32df5.png)

M is the magnetization saturation with respect to c1, c2, and c3. (Little bit of context here, I'm using the phase-field method to model the spinodal decomposition of FeCrCo.)

I modified the CoefDiffusion kernel to calculate ∇^2φ = x*y and compared the results to a benchmark from COMSOL with accurate results:

![image](https://user-images.githubusercontent.com/74686859/123172173-f98e2980-d439-11eb-949d-c33245703ef4.png)

My goal is to customize my custom kernel so that it can take M as a function from the Materials system and then take the gradient with respect to x (assuming that the magnetic direction is going towards the x direction). Here are my .C and .h files for the custom kernel:

```
#include "NewCoefDiffusion.h"

registerMooseObject("diffusion", NewCoefDiffusion);

InputParameters
NewCoefDiffusion::validParams()
{
  InputParameters params = Kernel::validParams();
  params.addParam<Real>("coef", 0.0, "Diffusion coefficient");
  params.addParam<FunctionName>("function",
                                "If provided, the diffusion coefficient will be coef + "
                                "this function.  This is useful for temporally or "
                                "spatially varying diffusivities");
  params.addClassDescription("Kernel for diffusion with diffusivity = coef + function");
  params.addRequiredParam<MaterialPropertyName>("m", "Base name of the Magnetization Saturation defined in a DertivativeParsedMaterial");
  return params;
}

NewCoefDiffusion::NewCoefDiffusion(const InputParameters & parameters)
  : Kernel(parameters),
    _coef(getParam<Real>("coef")),
    _func(parameters.isParamValid("function") ? &getFunction("function") : NULL),
    _m(getMaterialPropertyDerivative<Real>("m", _var.name()))
{
}

Real
NewCoefDiffusion::computeQpResidual()
{
  Real diffusivity = _coef;

  if (_func)
    diffusivity += _func->value(_t, _q_point[_qp]);

  return _grad_test[_i][_qp] * -_grad_u[_qp] - _m*grad_test[_i][_qp];
}

Real
NewCoefDiffusion::computeQpJacobian()
{
  Real diffusivity = _coef;

  if (_func)
    diffusivity += _func->value(_t, _q_point[_qp]);

  return _grad_test[_i][_qp] * -_grad_phi[_j][_qp] - _m*grad_test[_i][_qp];
}


#pragma once

#include "Kernel.h"
#include "Function.h"
#include "DerivativeMaterialInterface.h"

class NewCoefDiffusion : public Kernel
{
public:
  static InputParameters validParams();

  NewCoefDiffusion(const InputParameters & parameters);

protected:
  virtual Real computeQpResidual();
  virtual Real computeQpJacobian();

private:
  const Real _coef;
  const Function * const _func;
  const MaterialProperty<Real> & _m;
};

```
I thought following SplitCHParsed.C and SplitCHWRes.C and their implementation of Material properties would be straightforward, but it turns out it's not.

Any help would be appreciated! Thank you!

**Comment by ngrilli** [comment, marked as answer]
Dear @alanchavez86 

You can see an example in this kernel in my repository:
https://github.com/ngrilli/c_pfor_am/blob/main/src/kernels/ConservativeAdvectionCoupled.C

I get a material property (a vector) in the kernel using:
_edge_slip_direction(getMaterialProperty<std::vector<Real>>("edge_slip_direction")),

This material property is declared in this material:
https://github.com/ngrilli/c_pfor_am/blob/main/src/materials/FiniteStrainCrystalPlasticityDislo.C

as:
_edge_slip_direction(declareProperty<std::vector<Real>>("edge_slip_direction")),

and it is calculated inside the material object.

Trust this helps,
Nicolò Grilli
University of Bristol

**Reply by dschwen** [reply]
Hey Nicolo, some unsolicited comments: in the example you are using a _ton_ of `addCoupledVar`. Are you aware that we support vector coupling? I.e. the user can specify multiple coupled variable per `addCoupledVar` parameter. You can access those with the optional component parameter in `coupledValue("name", component)` and get the number of coupled components with `coupledComponents("name")`. The slip direction property should be of type `std::vector<RealVectorValue>` and be sized to `_nss` rather than `_nss * LIBMESH_DIM`.

**Reply by ngrilli** [reply]
Dear @dschwen 

Your comment is very useful to improve my code.
I was not aware of the coupled vectors. 
I was actually looking for that some time ago but I am not enough familiar with the framework.

Could you point me to an example kernel where vector coupling is implemented?

My idea is to develop a continuum dislocation dynamics code in which dislocation densities are variables
because I need to introduce the flux terms, dislocation multiplication, et cetera.
Then I couple the dislocation density variables with the crystal plasticity material 
by implementing Orowan law and calculating dislocation velocity. 

Indeed, if I could do vector variable coupling, I could develop a general code with an arbitrary number of active slip system,
while now the number of slip systems is fixed.

I am using the slip direction vector to pass the rotated slip directions of all slip systems
to the Advection kernel, so that I can implement dislocation flux along arbitrary directions on slip plane in 3D
depending on the crystal rotation matrix.

Please always write these unsolicited comments.

Nicolò

**Reply by alanchavez86** [reply]
@ngrilli 

Thank you for the suggestion! I went through and followed your kernel from your repository and modified my kernel. Looking into your files, I've noticed that there is a src/materials directory. Would I have to create a new materials directory and possibly add a custom material file?

Thank you.

**Reply by dschwen** [reply]
@ngrilli , here https://github.com/idaholab/moose/blob/06bc1f73f0d4dfb3d01ce9ba77d6bb97346e07eb/modules/phase_field/src/ics/CoupledValueFunctionIC.C is an example

https://github.com/idaholab/moose/blob/06bc1f73f0d4dfb3d01ce9ba77d6bb97346e07eb/modules/phase_field/include/ics/CoupledValueFunctionIC.h#L28

https://github.com/idaholab/moose/blob/06bc1f73f0d4dfb3d01ce9ba77d6bb97346e07eb/modules/phase_field/src/ics/CoupledValueFunctionIC.C#L22-L24

https://github.com/idaholab/moose/blob/06bc1f73f0d4dfb3d01ce9ba77d6bb97346e07eb/modules/phase_field/src/ics/CoupledValueFunctionIC.C#L32

Instead of `coupledValue("name", component)` I'm using the even simpler `coupledValues("name")` (note the plural) which gets a vector of variable pointers.

**Reply by ngrilli** [reply]
Dear @alanchavez86 

I keep the same structure as the MOOSE framework directory system,
therefore I include material classes in a /materials/ folder.

I am not sure if that is compulsory, but I don't see why you should not do that.
It's useful for distinguishing material classes from kernel classes.

Nicolò

**Reply by alanchavez86** [reply]
@ngrilli 

Thank you for the clarification. I wasn't aware that you can create your own material classes. I'm still fairly new at creating custom kernels so I didn't think of this at first. I'll try the parts of your code that you mentioned and let you know how it goes!

**Comment by dschwen** [comment]
At the very least you need to replace `_m` with `_m[_qp]`.

**Reply by dschwen** [reply]
Also you calculate `diffusivity` and never use it.

**Reply by alanchavez86** [reply]
@dschwen

Thank you for catching those. I must have overlooked those mistakes when modifying the CoefDIffusion kernel.
